src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+      async (resolvedCapabilties, modifier) => modifier(this.request, await resolvedCapabilties),
+      Promise.resolve(capabilities)
+    );
+  }
+
+  private async getVariables(app: App, basePath: string, injectedOverrides: Record<string, any>) {
+    const providers: Provider[] = (this.legacy.uiExports as any).defaultInjectedVarProviders || [];
+    const replacers: Replacer[] = (this.legacy.uiExports as any).injectedVarsReplacers || [];
+    const defaultInjectedVars = providers.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(
+          defaults,
+          fn(this.http.server, pluginSpec.readConfigValue(this.http.server.config, []))
+        ),
+      {}
+    );
+    const appInjectedVars = await this.http.server.getInjectedUiAppVars(app.getId());
+    const injectedVars = mergeVariables(defaultInjectedVars, appInjectedVars, injectedOverrides);
+    const vars = await replacers.reduce(
+      async (variables, replacer) => await replacer(variables, this.request, this.http.server),
+      Promise.resolve(injectedVars)
+    );
+
+    return {
+      strictCsp: this.http.csp.strict,
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${app.getId()}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: this.uiSettings?.user?.['theme:darkMode']?.userValue ?? false,
+      injectedMetadata: {
+        version: this.http.server.version,
+        buildNumber: this.env.packageInfo.buildNum,
+        branch: this.env.packageInfo.branch,
+        basePath,
+        env: this.env,
+        legacyMode: app.getId() !== 'core',
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: {
+          warnLegacyBrowsers: this.http.csp.warnLegacyBrowsers,
+        },
+        vars,
+        uiPlugins: [...this.plugins.uiPlugins.public].map(([id, plugin]) => ({ id, plugin })),
+        legacyMetadata: {
+          app,
+          bundleId: `app:${app.getId()}`,
+          nav: this.http.server.getUiNavLinks(),
+          env: this.env,
+          version: this.env.packageInfo.version,
+          branch: this.env.packageInfo.branch,
+          buildNum: this.env.packageInfo.buildNum,
+          buildSha: this.env.packageInfo.buildSha,
+          serverName: this.http.server.name,
+          devMode: this.env.mode.dev,
+          basePath,
+          uiSettings: this.uiSettings,
+        },
+        capabilities: await this.getCapabilities(),
+      },
+    };
+  }
+
+  public async render(appId: string, injectedVarsOverrides: Record<string, any> = {}) {
+    try {
+      const app = this.http.server.getUiAppById(appId) || { getId: () => 'core' };
+      const variables = await this.getVariables(
+        app,
+        this.http.basePath.get(this.request),
+        injectedVarsOverrides
+      );
+      const content = this.template(variables);
+
+      return this.response.ok({
```

Comment:
I'm not sure that we should couple rendering & http services. Can we render a report or email in a background task? I haven't given it a lot of thoughts, but I can imagine something like:
```js
// my-plugin/server/plugins.ts
return http.response.ok(context.renderService.app().toHtml())
```

Reply:
I think this is something we could explore in the future, but is out of scope for this patch. :+1: